### ISSUE TITLE
Bug: no-use-before-define does not treat IIFE bodies as the same execution context

### ISSUE THREAD

**Issue by irfanfandi** [eslint/eslint#21224]
### Environment

Node version: v22.22.3
npm version: v10.9.8
Local ESLint version: v10.8.1
Global ESLint version: not installed
Operating System: macOS 26.1

### What parser are you using?

Default (Espree)

### What did you do?

<details><summary>Configuration</summary>

```js
export default [
    {
        rules: {
            "no-use-before-define": ["error", { "variables": false }]
        }
    }
];
```

</details>

```js
/*eslint no-use-before-define: ["error", { "variables": false }]*/

(() => {
    console.log(a);
})();

const a = 5;
```

### What did you expect to happen?

An error on `a`, because the IIFE body runs immediately — before `a` is initialized — so this throws `ReferenceError: Cannot access 'a' before initialization` at runtime.

The rule already handles the analogous case for class static initializer blocks, which are also provably executed immediately. That case is listed as **incorrect** code for `{ "variables": false }` in the [rule docs](https://eslint.org/docs/latest/rules/no-use-before-define#variables):

```js
/*eslint no-use-before-define: ["error", { "variables": false }]*/

{
    class C {
        static {
            this.x = foo; // reported
        }
    }
    const foo = 1;
}
```

I'd expect an IIFE body to be treated the same way, since its immediacy is equally provable at the call site.

### What actually happened?

No error is reported. Verified on ESLint v10.8.1:

```
$ npx eslint test.js
(no output)

$ node test.js
ReferenceError: Cannot access 'a' before initialization
```

This affects both `{ "variables": false }` and `{ "classes": false }`, and every IIFE shape I tried:

```js
/*eslint no-use-before-define: ["error", { "variables": false }]*/

// 1. arrow IIFE — not reported
(() => { console.log(a); })();

// 2. function expression IIFE — not reported
(function () { console.log(a); })();

// 3. awaited async IIFE — not reported
await (async () => { console.log(a); })();

// 4. nested IIFEs — not reported
(() => { (() => { console.log(a); })(); })();

const a = 5;
```

```js
/*eslint no-use-before-define: ["error", { "classes": false }]*/

// 5. class reference inside an IIFE — not reported
(() => { new C(); })();

class C {}
```

For contrast, these are correctly **not** reported today and must stay that way:

```js
/*eslint no-use-before-define: ["error", { "variables": false }]*/

// deferred — the function isn't called until after `a` is initialized
const f = () => { console.log(a); };
const a = 5;
f();
```

```js
/*eslint no-use-before-define: ["error", { "variables": false }]*/

// the IIFE runs immediately, but the returned inner function is deferred
(() => () => a)();
const a = 5;
```

Note that with the default options (`variables: true`) case 1 *is* reported, so this only affects configurations that opt out via `variables: false` / `classes: false`.

### Link to Minimal Reproducible Example

https://eslint.org/play/#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tdXNlLWJlZm9yZS1kZWZpbmU6IFtcImVycm9yXCIsIHsgXCJ2YXJpYWJsZXNcIjogZmFsc2UgfV0qL1xuXG4oKCkgPT4ge1xuICAgIGNvbnNvbGUubG9nKGEpO1xufSkoKTtcblxuY29uc3QgYSA9IDU7XG4iLCJvcHRpb25zIjp7InJ1bGVzIjp7Im5vLXVzZS1iZWZvcmUtZGVmaW5lIjpbImVycm9yIix7InZhcmlhYmxlcyI6ZmFsc2V9XX0sImxhbmd1YWdlT3B0aW9ucyI6eyJzb3VyY2VUeXBlIjoibW9kdWxlIiwicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6ImxhdGVzdCJ9fX0=

### Participation

- [X] I am willing to submit a pull request for this issue.

### AI acknowledgment

- [ ] I did not use AI to generate this issue report.
- [ ] (If the above is not checked) I have reviewed the AI-generated content before submitting.

### Additional comments

This is the follow-up issue that @mdjermanovic asked for in #20014, so that IIFE handling can be tracked separately from the initializer question discussed there:

> I think treating code in IIFEs as code that runs in the upper execution context is not directly related to https://github.com/eslint/eslint/issues/20014 which is, I believe, about referencing variable in its initializer. […] Which looks fine to me, but I think it should be a separate bug report / feature request.
>
> — https://github.com/eslint/eslint/pull/20019#issuecomment-3215340614

> @kirkwaiblinger would you like to open a separate issue about handling IIFEs in this rule? It doesn't seem directly related to this issue. Also, I think it shouldn't be limited to variable initializers only […]
>
> — https://github.com/eslint/eslint/issues/20014#issuecomment-3448442713

**Scope note:** this issue is deliberately only about IIFEs *outside* of a variable's own initializer, so it doesn't overlap with #20014 / #20248. The self-referential initializer case (`const b = (() => b)();`) is already covered there.

**Possible approach.** `isFromSeparateExecutionContext()` ([`lib/rules/no-use-before-define.js#L132-L147`](https://github.com/eslint/eslint/blob/main/lib/rules/no-use-before-define.js#L132-L147)) already walks up out of class static initializer scopes. An IIFE scope could be treated the same way, as @kirkwaiblinger prototyped in the closed draft #20019:

```js
function isIife(scope) {
    return (
        scope.type === "function" &&
        scope.block.parent.type === "CallExpression" &&
        scope.block.parent.callee === scope.block
    );
}
```

That already gives the right answer for the deferred counter-examples above, because the walk stops at the first scope that isn't immediately invoked.

Two things I'd like guidance on before opening a PR:

1. **Scope of "immediately invoked".** Should `.call()` / `.apply()` forms be included? They're equally provable but not syntactically IIFEs:

   ```js
   (function () { console.log(a); }).call(null); // not reported today
   const a = 5;
   ```

2. **Breaking-ness.** This only changes behavior when `variables: false` or `classes: false` is set, and only turns false negatives into true positives, so I'd expect it to be a normal bug fix rather than a breaking change — but happy to put it behind an option if the team prefers.

I'm happy to implement whichever direction the team decides.

**Comment by mdjermanovic** [comment]
> ```js
> /*eslint no-use-before-define: ["error", { "variables": false }]*/
> 
> (() => {
>     console.log(a);
> })();
> 
> const a = 5;
> ```

I'm in favor of reporting these cases by default (no option needed) and treating this change as a semver-minor bug fix. Would like more opinions from @eslint/eslint-tsc.

**Comment by seongmin36** [comment]
I've opened #21237  implementing this. review would be much appreciated!

**Comment by seongmin36** [comment]
Hi @irfanfandi sorry for the noise, I opened a PR here before noticing it
was assigned to you.

One finding in case it's useful: the change also affects the default options,
since `const b = (() => b)();` reaches the check through
`isEvaluatedDuringInitialization()` rather than the options gate — might
matter for the semver-minor question.

**Comment by fasttime** [comment]
> > ```js
> > /*eslint no-use-before-define: ["error", { "variables": false }]*/
> > 
> > (() => {
> >     console.log(a);
> > })();
> > 
> > const a = 5;
> > ```
> 
> I'm in favor of reporting these cases by default (no option needed) and treating this change as a semver-minor bug fix. Would like more opinions from [@eslint/eslint-tsc](https://github.com/orgs/eslint/teams/eslint-tsc).

I'm also in favor of reporting those cases and treating the change as non-breaking.

Noting that it would be good to revisit the suggested approach to account for cases where the IIFE execution can be delayed, for example with async functions (not uncommon in script code at top-level), or generators.

**Comment by irfanfandi** [comment]
Thanks @seongmin36 for checking, and no worries about the overlap!

Since this is assigned to me, I'd like to go ahead with the PR — I already have an implementation ready. It also addresses the generator/async gap @fasttime raised: a generator IIFE like `(function* () { console.log(a); })().next()` doesn't actually run its body at call time, only on the first `.next()`, so it shouldn't be treated as immediate. `isIIFEScope()` needs to exclude generator functions for correctness (and I'm double-checking the async case too, since execution can suspend at an `await`).

Will open the PR shortly.

**Comment by seongmin36** [comment]
Thanks for the update, @irfanfandi! That makes total sense, and I look forward to seeing your PR. Good luck with the implementation!

**Comment by lumirlumir** [comment]
As there are two approvals from the TSC, I’m marking this issue as accepted.

**Comment by scenerecap** [comment]
Reproduced on eslint.org/play (ESLint v9.28.0, 2026-08-25, Chromium, 1280×720):

1. Configure `no-use-before-define: ["error", { "variables": false }]`.
2. Enter the IIFE snippet:
   ```js
   (() => {
     console.log(a);
   })();
   const a = 5;
   ```
3. Output panel: no errors reported.
4. Control — replace the IIFE with a class static initialiser block (`class C { static { this.x = foo; } } const foo = 1;`): the rule correctly fires.

The IIFE runs immediately, so `a` is accessed before its `const` binding, but the rule treats the IIFE body as a lazy scope and skips it. 24-second recording with both cases side by side: https://scenerecap.com/s/5kzxakpt7i

(#21238 is open and addresses this.)
